flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSink.java

```diff
@@ -0,0 +1,255 @@
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
+import org.apache.flink.streaming.api.functions.sink.RichSinkFunction;
+import org.apache.flink.streaming.api.functions.sink.SinkFunction;
+
+import com.google.api.gax.core.FixedCredentialsProvider;
+import com.google.api.gax.core.NoCredentialsProvider;
+import com.google.api.gax.grpc.GrpcTransportChannel;
+import com.google.api.gax.rpc.FixedTransportChannelProvider;
+import com.google.api.gax.rpc.TransportChannel;
+import com.google.auth.Credentials;
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.cloud.pubsub.v1.SubscriptionAdminSettings.defaultCredentialsProviderBuilder;
+
+/**
+ * A sink function that outputs to PubSub.
+ *
+ * @param <IN> type of PubSubSink messages to write
+ */
+public class PubSubSink<IN> extends RichSinkFunction<IN> implements CheckpointedFunction {
+	private static final Logger LOG = LoggerFactory.getLogger(PubSubSink.class);
+
+	private final Credentials credentials;
+	private final SerializationSchema<IN> serializationSchema;
+	private final String projectName;
+	private final String topicName;
+	private final String hostAndPortForEmulator;
+
+	private transient Publisher publisher;
+
+	private PubSubSink(
+		Credentials credentials,
+		SerializationSchema<IN> serializationSchema,
+		String projectName,
+		String topicName,
+		String hostAndPortForEmulator) {
+		this.credentials = credentials;
+		this.serializationSchema = serializationSchema;
+		this.projectName = projectName;
+		this.topicName = topicName;
+		this.hostAndPortForEmulator = hostAndPortForEmulator;
+	}
+
+	private transient ManagedChannel managedChannel = null;
+	private transient TransportChannel channel = null;
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		Publisher.Builder builder = Publisher
+			.newBuilder(ProjectTopicName.of(projectName, topicName))
+			.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+
+		if (hostAndPortForEmulator != null) {
+			managedChannel = ManagedChannelBuilder
+				.forTarget(hostAndPortForEmulator)
+				.usePlaintext(true) // This is 'Ok' because this is ONLY used for testing.
+				.build();
+			channel = GrpcTransportChannel.newBuilder().setManagedChannel(managedChannel).build();
+			builder.setChannelProvider(FixedTransportChannelProvider.create(channel))
+					.setCredentialsProvider(NoCredentialsProvider.create());
+		}
+
+		publisher = builder.build();
+	}
+
+	@Override
+	public void close() throws Exception {
```

Comment:
Same as the source, is there a thread-safety issue here?

Reply:
See my comment in the source